### ISSUE TITLE
Runner Pods randomly receiving a sigterm and killing jobs

### ISSUE THREAD

**Issue by maeghan-porter** [actions/actions-runner-controller#2695]
### Checks

- [X] I've already read https://github.com/actions/actions-runner-controller/blob/master/TROUBLESHOOTING.md and I'm sure my issue is not covered in the troubleshooting guide.
- [X] I'm not using a custom entrypoint in my runner image

### Controller Version

0.27.3

### Helm Chart Version

0.23.2

### CertManager Version

1.10.0

### Deployment Method

Helm

### cert-manager installation

I am installing it via helm as outlined here: https://cert-manager.io/docs/installation/helm/
With the `installCRDs=true` option set.

### Checks

- [X] This isn't a question or user support case (For Q&A and community support, go to [Discussions](https://github.com/actions/actions-runner-controller/discussions). It might also be a good idea to contract with any of contributors and maintainers if your business is so critical and therefore you need priority support
- [X] I've read [releasenotes](https://github.com/actions/actions-runner-controller/tree/master/docs/releasenotes) before submitting this issue and I'm sure it's not due to any recently-introduced backward-incompatible changes
- [X] My actions-runner-controller version (v0.x.y) does support the feature
- [X] I've already upgraded ARC (including the CRDs, see charts/actions-runner-controller/docs/UPGRADING.md for details) to the latest and it didn't fix the issue
- [X] I've migrated to the workflow job webhook event (if you using webhook driven scaling)

### Resource Definitions

```yaml
apiVersion: actions.summerwind.dev/v1alpha1
kind: RunnerDeployment
metadata:
  name: runner-deployment-nonprod
  namespace: gha
spec:
  template:
    spec:
      image: docker-all.repo.sonatype.com/summerwind/actions-runner-dind:v2.303.0-ubuntu-22.04
      imagePullSecrets:
        - name: repo.sonatype.com
      dockerdWithinRunnerContainer: true
      organization: sonatype
      labels:
        - stable
        - 0-8-0
        - nonprod
      group: my-runner-group
---
apiVersion: actions.summerwind.dev/v1alpha1
kind: HorizontalRunnerAutoscaler
metadata:
  name: runner-autoscaler-nonprod
  namespace: gha
spec:
  scaleDownDelaySecondsAfterScaleOut: 300
  scaleTargetRef:
    kind: RunnerDeployment
    name: runner-deployment-nonprod
  minReplicas: 2
  maxReplicas: 10
  metrics:
    - type: PercentageRunnersBusy
      scaleUpThreshold: '0.75'
      scaleDownThreshold: '0.25'
      scaleUpFactor: '2'
      scaleDownFactor: '1'
```

### To Reproduce

```markdown
I can't really reproduce it on demand, it happens randomly.
```

### Describe the bug

It is described in this question from March: https://github.com/actions/actions-runner-controller/discussions/2417
But no one has said anything on it and since multiple people are experiencing it I am filing it as an issue/bug.

The TL;DR is that runner pods are randomly getting a shutdown signal for no reason that anyone can gather from the logs or from the metrics of the instances at the time of the shutdown signal and it cancels whatever job it is running at the moment.

I am running 8vCore 32GB memory instances and there were no significant resource use increases at the time of the sigterm. 

### Describe the expected behavior

I expect that jobs shouldn't be cancelled randomly for no obvious reason.

### Whole Controller Logs

```shell
https://gist.github.com/maeghan-porter/aceb7352951cd0a834dc133051772f9f
```

### Whole Runner Pod Logs

```shell
https://gist.github.com/maeghan-porter/88973c0dbd24ab55d4162a5cc037ffff
```

**Comment by mumoshu** [comment]
Hey @maeghan-porter!
I believe it's due to you're using spot instances. Are you sure you've tweaked all your workflow jobs to gracefully stop/finish within 2 minutes(which is the max available time until the spot instance is terminated after the spot termination notice) AND you've configured your cluster correctly to let the node termination waits until the pod termination?

Do note that the node termination and therefore the termination of all the runner pods on the node should complete within at most 2 minutes.

Otherwise your workflow jobs will "hang" for some time until GitHub Actions finally considers the job/runner has disappeared.

To be clear, ARC has no way to prevent spot instances from terminating and I presume that's the root cause of your issue.

**Comment by Langleu** [comment]
Hey @maeghan-porter, just to add another possible explanation.
I've noticed that we had too little disk space available, which resulted in the same error.
Like 7-10% were left on the machine (out of a 100GB) and the error happened as well ... every single time.
We were running some postgres dump restore and from looking at it, there was still plenty left but when we increased the disk space, the problem was resolved.
Always thought that 10% would be enough but seems not to be the case.
I never followed up on the issue but would assume not a controller issue but rather Kubernetes / agent related.

But maybe worth looking into that whether you're potentially running into the same.

**Comment by maeghan-porter** [comment]
> Hey @maeghan-porter! I believe it's due to you're using spot instances. Are you sure you've tweaked all your workflow jobs to gracefully stop/finish within 2 minutes(which is the max available time until the spot instance is terminated after the spot termination notice) AND you've configured your cluster correctly to let the node termination waits until the pod termination?
> 
> Do note that the node termination and therefore the termination of all the runner pods on the node should complete within at most 2 minutes.
> 
> Otherwise your workflow jobs will "hang" for some time until GitHub Actions finally considers the job/runner has disappeared.
> 
> To be clear, ARC has no way to prevent spot instances from terminating and I presume that's the root cause of your issue.

Thanks for your reply! 
I was wondering if it was spot. Do you have any docs on hand on how to "handle" this properly in workflows? It is worth noting that we have some rather big terraform jobs that this is very bad for (building eks clusters actually), and I'm not sure if it's really possible to handle that situation gracefully in 2 minutes. We use the aws node termination handler in the cluster, if that's relevant.

I'm thinking we should switch to on-demand at this point.

**Comment by maeghan-porter** [comment]
> Hey @maeghan-porter, just to add another possible explanation. I've noticed that we had too little disk space available, which resulted in the same error. Like 7-10% were left on the machine (out of a 100GB) and the error happened as well ... every single time. We were running some postgres dump restore and from looking at it, there was still plenty left but when we increased the disk space, the problem was resolved. Always thought that 10% would be enough but seems not to be the case. I never followed up on the issue but would assume not a controller issue but rather Kubernetes / agent related.
> 
> But maybe worth looking into that whether you're potentially running into the same.

Thanks for your reply! I am having space issues AS WELL, but those actually explicitly have errors referring to no storage space being the issue. This issue here does not indicate that. That doesn't mean it isn't possible though! I would expect the error message to be consistent though 🤔

**Comment by Langleu** [comment]
It's not controller related.
If you catch the runner pod that's being evicted, you may find an event like the following
```
  Warning  Evicted              44s   kubelet  The node was low on resource: ephemeral-storage. Container runner was using 3914072Ki, which exceeds its request of 0. Container docker was using 91249008Ki, which exceeds its request of 0.
```

Eviction --> `Error: The operation was canceled.`.
Maybe your issue is a different one, but for us it was reproducible that the runner pod got evicted for using up too much storage of the node resulting in exactly the same error.

**Comment by maeghan-porter** [comment]
> It's not controller related. If you catch the runner pod that's being evicted, you may find an event like the following
> 
> ```
>   Warning  Evicted              44s   kubelet  The node was low on resource: ephemeral-storage. Container runner was using 3914072Ki, which exceeds its request of 0. Container docker was using 91249008Ki, which exceeds its request of 0.
> ```
> 
> Eviction --> `Error: The operation was canceled.`. Maybe your issue is a different one, but for us it was reproducible that the runner pod got evicted for using up too much storage of the node resulting in exactly the same error.

Hmm I haven't noticed anything like that but I suppose it could be easy to miss.

**Comment by maeghan-porter** [comment]
i'm trying to deal with the storage issues, which it seems this was caused by, though the inconsistent errors aren't great. But I'm closing this now. Thanks for the assistance, all!

**Comment by MPV** [comment]
Related to this upstream issue?
- https://github.com/actions/runner/issues/2582

**Comment by maeghan-porter** [comment]
> Related to this upstream issue?
> 
> * [The linux runner does not gracefully shutdown on SIGINT runner#2582](https://github.com/actions/runner/issues/2582)

I don't believe so. I am using ephemeral runners, and according to this issue they are not and apparently ephemeral runners would solve this for them.
